### PR TITLE
Feature/share intent

### PR DESCRIPTION
**Pull request by sbellity**

### CODE REVIEW

File: src/client/sharer/whatsapp.coffee [hunk 1]
```diff
@@ -1,20 +0,0 @@
...
-GenericShare = require './generic_share'
-
-class WhatsappShare extends GenericShare
-  constructor : (api, auth, currentUser, opts)->
-    super(api, currentUser, 'facebook')
-
-    @opts = opts
-    @params = opts.params || {}
-
-    t = "#{document.title} #{@params.url}"
```
Comment (unity) [inline src/client/sharer/whatsapp.coffee]: Smart defaults like this are really useful. Have we kept them ?

File: src/client/sharer/index.coffee [hunk 2]
```diff
@@ -108,32 +48,29 @@ class Sharer
     # 2. Find url from Click Targt
     # 3. Ship container node
 
-    opts.params.url ||= findUrl(event.target)
+    opts.params.url = opts.params.url || opts.params.href
+
+    if (!opts.params.url)
+      opts.params.url = findUrl(event.target)
+      opts.params.title = opts.params.title || domWalker.getMetaValue('og:title') || document.title
```
Comment (oliger) [inline src/client/sharer/index.coffee]: When sharing an item from the homepage of a website `Hull.share({ provider: 'twitter': url: 'http://acme.com/item/123' })`, the text in the twitter sharer is the title of the homepage not the title of the item.

To me if `params.title` is `null` we should fetch the title of the provided url server side.

File: src/client/sharer/index.coffee [hunk 3]
```diff
@@ -1,94 +1,34 @@
...
+  querystring = to_qs(params)
+  share = window.open("#{location}?#{querystring}", 'hull_share', "location=0,status=0,width=#{opts.width},height=#{opts.height}")
+  new Promise (resolve, reject)->
+    interval = setInterval ()->
+      try
+        if share == null || share.closed
+          window.clearInterval(interval)
+          resolve({display:"popup"})
+       catch e
+        1 == 1
```
Comment (oliger) [inline src/client/sharer/index.coffee]: not sure to understand `1 == 1`.

File: src/client/sharer/index.coffee [hunk 4]
```diff
@@ -1,94 +1,34 @@
...
+    encodeURIComponent(k)+'='+encodeURIComponent(v)
+  .join('&')
+
+popup = (location, opts={}, params={})->
+  return unless location?
+  querystring = to_qs(params)
+  share = window.open("#{location}?#{querystring}", 'hull_share', "location=0,status=0,width=#{opts.width},height=#{opts.height}")
+  new Promise (resolve, reject)->
+    interval = setInterval ()->
+      try
```
Comment (oliger) [inline src/client/sharer/index.coffee]: Do we need a `try/catch`?

File: src/client/sharer/index.coffee [hunk 5]
```diff
@@ -1,94 +1,34 @@
...
-    if id
-      ['utm_content', id]
-    else
-      []
-
-}
+assign      = require '../../polyfills/assign'
+domWalker = require '../../utils/dom-walker'
+
+to_qs = (params)->
```
Comment (oliger) [inline src/client/sharer/index.coffee]: https://github.com/hull/hull-js/blob/release/0.9.0/src/utils/query-string-encoder.coffee ?

File: src/client/sharer/index.coffee [hunk 6]
```diff
@@ -108,32 +48,29 @@ class Sharer
...
 
-    sharePromise = switch opts.provider
-      when 'email'    then new EmailShare(@api, @auth, @currentUser, opts, @config)
-      when 'facebook' then new FacebookShare(@api, @auth, @currentUser, opts, @config)
-      when 'twitter'  then new TwitterShare(@api, @auth, @currentUser, opts, @config)
-      when 'google'   then new GoogleShare(@api, @auth, @currentUser, opts, @config)
-      when 'linkedin' then new LinkedinShare(@api, @auth, @currentUser, opts, @config)
-      when 'whatsapp' then new WhatsappShare(@api, @auth, @currentUser, opts, @config)
+    # debugger
+    # return
```
Comment (oliger) [inline src/client/sharer/index.coffee]: ;)

File: src/client/sharer/index.coffee (hunk 4 above)
Comment (sbellity) [inline src/client/sharer/index.coffee]: I'm not sure, it was already there... @unity is there a good reason to have it ?

File: src/client/sharer/index.coffee (hunk 3 above)
Comment (sbellity) [inline src/client/sharer/index.coffee]: https://github.com/hull/hull-js/blob/09c43d345b08b56659ccc16dab4e3ac504ea16a3/src/client/sharer/generic_share.coffee#L31

**Comment by oliger** [comment]
> When sharing an item from the homepage of a website Hull.share({ provider: 'twitter': url: 'http://acme.com/item/123' }), the text in the twitter sharer is the title of the homepage not the title of the item.
>
> To me if params.title is null we should fetch the title of the provided url server side.

@sbellity Hi, what is the behavior now?

**Comment by sbellity** [comment]
@oliger this needs to be managed server side in the share endpoint. 
I have not added it yet because it bugs me to have an endpoint with synchronous http calls like that. Need to think about the best way to add this feature.

**Comment by oliger** [comment]
ok ok

**Pull request merged**